File: Algorithm/9/knapsack01.py
class Solution(object):
    def __init__(self):
        self.memo = []

    def knapsack01(self,w, v, C):
        n = len(w)
        self.memo = [[-1 for i in range(C+1)] for i in range(n)]
        return self.bestValue(w, v, n-1, C);

    def bestValue(self, w, v, index, c):
        if index < 0 or c <= 0:
            return 0
        if self.memo[index][c] != -1:
            return self.memo[index][c]

        res = self.bestValue(w,v,index-1,c)
        if c >= w[index]:
            res = max(res, v[index] + self.bestValue(w, v, index-1, c-w[index]))
        self.memo[index][c] = res;
        return res;

class Solution1(object):
    def knapsack01(self,w,v,C):
        n = len(w)
        if n == 0:
            return 0

        memo = [[-1 for i in range(C+1)] for i in range(n)]

        for j in range(C+1):
            if j >= w[0]:
                memo[0][j] = v[0]
            else:
                memo[0][j] = 0


        for i in range(1,n):
            for j in range(C+1):
                memo[i][j] = memo[i-1][j]
                if j >= w[i]:
                    memo[i][j] = max(memo[i][j], v[i] + memo[i-1][j-w[i]])

        return memo[n-1][C]

class Solution2(object):
    def knapsack01(self,w,v,C):
        n = len(w)
        if n == 0:
            return 0

        memo = [[-1 for i in range(C+1)] for i in range(2)]

        for j in range(C+1):
            if j >= w[0]:
                memo[0][j] = v[0]
            else:
                memo[0][j] = 0

        for i in range(1,n):
            for j in range(C+1):
                memo[i%2][j] = memo[(i-1)%2][j]
                if j >= w[i]:
                    memo[i%2][j] = max(memo[i%2][j], v[i] + memo[(i-1)%2][j-w[i]])

        return memo[(n-1)%2][C]

class Solution3(object):
    def knapsack01(self,w,v,C):
        n = len(w)
        if n == 0:
            return 0

        memo = [-1 for i in range(C+1)]

        for j in range(C+1):
            if j >= w[0]:
                memo[j] = v[0]
            else:
                memo[j] = 0

        for i in range(1,n):
            for j in range(C, w[i]-1, -1):
                memo[j] = max(memo[j], v[i] + memo[j-w[i]])

        return memo[C]
if __name__ == '__main__':
    w = [1,2,3]
    v = [6,10,12]
    C = 5
    print(Solution1().knapsack01(w,v,C))

File: Algorithm/9/knapsack01.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

class Knapsack01{
private:
    vector<vector<int>> memo;

    // 用[0...index]的物品，填充容积为c的背包的最大价值
    int bestValue(const vector<int> &w, const vector<int> v, int index, int c){
        if(index < 0 || c <= 0)
            return 0;

        if(memo[index][c] != -1)
            return memo[index][c];

        int res = bestValue(w, v, index-1, c);
        if(c >= w[index])
            res = max(res, v[index] + bestValue(w, v, index-1, c-w[index]));
        memo[index][c] = res;
        return res;
    }
public:
    int knapsack01(const vector<int> &w, const vector<int> &v, int C){
        int n = w.size();
        memo = vector<vector<int>>(n, vector<int>(C+1, -1));
        return bestValue(w, v, n-1, C);
    }
};

class Knapsack01{
public:
    int knapsack01(const vector<int> &w, const vector<int> &v, int C){
        assert(w.size() == v.size());
        int n = w.size();
        if(n == 0)
            return 0;

        vector<vector<int>> memo(n, vector<int>(C+1, -1));

        for(int j = 0; j <= C; j++)
            memo[0][j] = (j >= w[0] ? v[0] : 0);

        for(int i = 1; i < n; i++)
            for(int j = 0; j <= C; j++){
                memo[i][j] = memo[i-1][j];
                if(j >= w[i])
                    memo[i][j] = max(memo[i][j], v[i] + memo[i-1][j-w[i]]);
            }

        return memo[n-1][C];
    }
};

class Knapsack01{
public:
    int knapsack01(const vector<int> &w, const vector<int> &v, int C){
        assert(w.size() == v.size());
        int n = w.size();
        if(n == 0)
            return 0;

        vector<vector<int>> memo(2, vector<int>(C+1, -1));

        for(int j = 0; j <= C; j++)
            memo[0][j] = (j >= w[0] ? v[0] : 0);

        for(int i = 1; i < n; i++)
            for(int j = 0; j <= C; j++){
                memo[i%2][j] = memo[(i-1)%2][j];
                if(j >= w[i])
                    memo[i%2][j] = max(memo[i%2][j], v[i] + memo[(i-1)%2][j-w[i]]);
            }

        return memo[(n-1)%2][C];
    }
};

class Knapsack01{
public:
    int knapsack01(const vector<int> &w, const vector<int> &v, int C){
        assert(w.size() == v.size());
        int n = w.size();
        if(n == 0)
            return 0;

        vector<int> memo(C+1, -1);

        for(int j = 0; j <= C; j++)
            memo[j] = (j >= w[0] ? v[0] : 0);

        for(int i = 1; i < n; i++)
            for(int j = C; j >= w[i]; j--)
                memo[j] = max(memo[j], v[i] + memo[j-w[i]]);

        return memo[C];
    }
};

int main(){

    return 0;
}
